src/main/java/org/kohsuke/stapler/idea/extension/JavaStructureViewExtension.java

```diff
@@ -0,0 +1,49 @@
+package org.kohsuke.stapler.idea.extension;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+import com.intellij.ide.structureView.StructureViewTreeElement;
+import com.intellij.openapi.vfs.VfsUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiClass;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import org.kohsuke.stapler.idea.psi.JellyFile;
+import org.kohsuke.stapler.idea.psi.link.LinkJellyFileImpl;
+import org.kohsuke.stapler.idea.structureview.LinkJellyFileTreeElement;
+
+/**
+ * Additional jelly files in java structure view
+ * Items are retrieved in related src/main/java directory
+ *
+ * @author Julien Greffe
+ */
+public class JavaStructureViewExtension extends AbstractStructureViewExtension {
+
+    @Override
+    public Class<? extends PsiElement> getType() {
+        return PsiClass.class;
+    }
+
+    @Override
+    public StructureViewTreeElement[] getChildren(PsiElement parent) {
+        final PsiManager psiManager = PsiManager.getInstance(parent.getProject());
+        final List<LinkJellyFileTreeElement> files = new ArrayList<>();
+        final Path expectedPath = getExpectedJellyParentPath(getJellyDirectoryPath(parent).toString());
+
+        VirtualFile containingJellyDirectory = VfsUtil.findFile(expectedPath, false);
+        if (containingJellyDirectory != null) {
+            for (VirtualFile jellyFile : containingJellyDirectory.getChildren()) {
+                PsiFile file = psiManager.findFile(jellyFile);
+                if (file instanceof JellyFile) {
+                    files.add(new LinkJellyFileTreeElement(
+                        new LinkJellyFileImpl(file.getViewProvider(), file.getFileElementType())));
+                }
+            }
+        }
+        return files.toArray(LinkJellyFileTreeElement.EMPTY_ARRAY);
+    }
```

Comment:
I started messing around with this locally to find a bit more "conceptual" alternative. I ended up with

```
String qualifiedClassName = ((PsiClass) parent).getQualifiedName();
        final List<StaplerViewTreeElement> files = new ArrayList<>();
        if (qualifiedClassName != null) {
            PsiPackage psiPackage = JavaPsiFacade.getInstance(parent.getProject())
                .findPackage(qualifiedClassName);
            if (psiPackage != null) {
                PsiFile[] maybeViewFiles = psiPackage.getFiles(GlobalSearchScope.moduleScope(ModuleUtil.findModuleForPsiElement(parent)));
                for (PsiFile file : maybeViewFiles) {
                    if (file instanceof JellyFile) {
                        files.add(new StaplerViewTreeElement((JellyFile) file));
                    }
                }
            }
        }
        return files.toArray(LinkJellyFileTreeElement.EMPTY_ARRAY);
```

The idea is that the views are in the Java Package that has the same fully qualified name as a Class. Hence, PSI objects are used to get a Java Package and then check the files inside the package but only in the same module as the Java Class. This seems to work. I have not checked if it works for classes in `jar` files, though.

A bit later I also remembered about the [GotoViewAction](https://github.com/jenkinsci/idea-stapler-plugin/blob/2.0.8/src/main/java/org/kohsuke/stapler/idea/GotoViewAction.java) wich is also doing view lookups based on `PsiClass`. It turns out it is also using the `PsiPackage` approach, although a bit differently.

----

You might also notice the `StaplerViewTreeElement`. It is a simple implementation of `StructureViewTreeElement` I implemented to avoid infinite recursion that you mentioned and the problem with Class' views showing up in the Structure View of Jelly file.

Reply:
Hello @duemir,
thanks for your input, it really simplifies the all behavior 👍🏼 
Applying this, and also updating the `JavaStructureViewExtension` to use same kind of utils.